netaddr.go

```diff
@@ -179,24 +179,55 @@ func MustParseIP(s string) IP {
 // net.ParseIP at the moment, so returning ok=false from here just
 // means it'll try it a slower way.
 func parseIPv4(s string) (ip IP, ok bool) {
-	if len(s) < 7 || len(s) > 15 || strings.Count(s, ".") != 3 {
-		return ip, false
-	}
-	var v [4]uint64
-	for i := range v {
-		var numStr string
-		if dot := strings.IndexByte(s, '.'); dot == -1 {
-			numStr = s
-		} else {
-			numStr, s = s[:dot], s[dot+1:]
+	// Parse directly into the return value. This shaves another ~5%
+	// off the parse speed compared to using a tmp array and the
+	// IPv4() constructor.
+	ip.lo = 0xffff
+	ip.z = z4
+
+	for i := 0; i < 4; i++ {
+		var (
+			j int
+			acc uint64
+		)
+		// Parse one byte of digits. Bail if we overflow, stop at
+		// first non-digit.
+		//
+		// As of Go 1.15, don't try to factor "parsedecimal" into a
+		// helper function. Its complexity is slightly too high for
+		// inlining, which ends up costing +50% in parse time.
+		for j = 0; j < len(s); j++ {
+			if s[j] < '0' || s[j] > '9' {
+				break
+			}
+			acc = (acc*10)+uint64(s[j]-'0')
+			if acc > 255 {
+				return IP{}, false
+			}
 		}
-		var err error
-		v[i], err = strconv.ParseUint(numStr, 10, 8)
-		if err != nil {
-			return ip, false
+		// There must be at least 1 digit per quad.
+		if j == 0 {
+			return IP{}, false
+		}
+
+		ip.lo = (ip.lo<<8)+acc
```

Comment:
or `|` instead of `+` but doubt it matters

Reply:
I tried both while implementing, no change in performance, so I kept the one that felt more obvious. That said, this code is gone now that the function uses IPv4().